subsys/net/lib/dhcpv6/dhcpv6.c

```diff
@@ -2226,3 +2295,77 @@ int net_dhcpv6_init(void)
 
 	return 0;
 }
+
+int net_dhcpv6_subnet_alloc(struct net_if *iface, struct in6_addr *prefix)
+{
+	struct net_if_dhcpv6 *dhcp = &iface->config.dhcpv6;
+	size_t max;
+	int ret;
+
+	if (!IS_ENABLED(CONFIG_NET_CONFIG_DHCPV6_PREFIX_DELEGATION)) {
+		return -ENOTSUP;
+	}
+
+	if (dhcp->params.request_prefix) {
+		return -EINVAL;
+	}
+
+	if (dhcp->state != NET_DHCPV6_BOUND) {
+		return -EINPROGRESS;
+	}
+
+	k_mutex_lock(&lock, K_FOREVER);
+
+	max = MIN(SUBNET_MASK_LEN, 1 << (IPV6_SUBNET_BITS - iface->config.dhcpv6.prefix_len));
+
+	for (size_t i = 0; i < max; ++i) {
+		if ((dhcp->delegated_subnets_mask & BIT(i)) == 0) {
+			memcpy(prefix, &iface->config.dhcpv6.prefix, sizeof(struct in6_addr));
+
+			dhcp->delegated_subnets_mask |= BIT(i);
+			prefix->s6_addr[SUBNET_ID_POS] |= (uint8_t)i;
+
+			ret = 0;
+
+			goto out;
+		}
+	}
+
+	ret = -ENOBUFS;
+
+out:
+	k_mutex_unlock(&lock);
+
+	return ret;
+}
+
+int net_dhcpv6_subnet_free(struct net_if *iface, struct in6_addr *prefix)
+{
+	struct net_if_dhcpv6 *dhcp = &iface->config.dhcpv6;
+	size_t mask_len;
+	uint8_t subnet_id;
+	int ret;
+
+	if (!IS_ENABLED(CONFIG_NET_CONFIG_DHCPV6_PREFIX_DELEGATION)) {
```

Comment:
Ditto, I suppose this could be DHCPv6-level option (https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/lib/dhcpv6/Kconfig)?